internal/frames.go

```diff
@@ -0,0 +1,141 @@
+/*
+   Copyright 2021 Gravitational, Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package internal
+
+import (
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// Trace stores structured trace entry, including file line and path
+type Trace struct {
+	// Path is a full file path
+	Path string `json:"path"`
+	// Func is a function name
+	Func string `json:"func"`
+	// Line is a code line number
+	Line int `json:"line"`
+}
+
+// FrameCursor stores the position in a call stack
+type FrameCursor struct {
+	// Current specifies the current stack frame.
+	// if omitted, rest contains the complete stack
+	Current *runtime.Frame
+	// Rest specifies the rest of stack frames to explore
+	Rest *runtime.Frames
+	// N specifies the total number of stack frames
+	N int
+}
+
+// Traces is a list of trace entries
+type Traces []Trace
+
+// CaptureTraces gets the current stack trace with some deep frames ignored
+func CaptureTraces(depth int) Traces {
+	var buf [32]uintptr
+	n := runtime.Callers(depth+2, buf[:])
+	pcs := buf[:n]
+	frames := runtime.CallersFrames(pcs)
+	cursor := FrameCursor{
+		Rest: frames,
+		N:    n,
+	}
+	return CaptureTracesFromCursor(cursor)
+}
+
+// CaptureTracesFromCursor gets the current stack trace from a given cursor
+func CaptureTracesFromCursor(cursor FrameCursor) Traces {
+	traces := make(Traces, 0, cursor.N)
+	if cursor.Current != nil {
+		traces = append(traces, frameToTrace(*cursor.Current))
+	}
+	for {
+		frame, more := cursor.Rest.Next()
+		traces = append(traces, frameToTrace(frame))
+		if !more {
+			break
+		}
+	}
+	return traces
+}
+
+func frameToTrace(frame runtime.Frame) Trace {
+	return Trace{
+		Func: frame.Function,
+		Path: frame.File,
+		Line: frame.Line,
+	}
+}
+
+// SetTraces adds new traces to the list
+func (s Traces) SetTraces(traces ...Trace) {
+	s = append(s, traces...)
+}
+
+// Func returns first function in trace list
+func (s Traces) Func() string {
+	if len(s) == 0 {
+		return ""
+	}
+	return s[0].Func
+}
+
+// Func returns just function name
+func (s Traces) FuncName() string {
+	if len(s) == 0 {
+		return ""
+	}
+	fn := filepath.ToSlash(s[0].Func)
+	idx := strings.LastIndex(fn, "/")
+	if idx == -1 || idx == len(fn)-1 {
```

Comment:
If `fn` ends with a trailing slash, the trailing slash will be returned as well here. Is this intended?

Reply:
I don't know the original motivation, I just moved the method from one file to another but... IMO it's a sort of redundant measure for preventing the `return` of empty string since empty string is kinda more ambiguous than a full path with slashes.  But in fact I think that a case with trailing slash is not actually possible.